### PR TITLE
Fix #1855: Multiassign from Union (take 2)

### LINKED ISSUE #1575: Type error unpacking Union of Lists
Version: master branch, 001ae51778baf3e2351d3efc63dc4eb4a191acef

Code:

``` python
from typing import Union

def f(x: bool) -> Union[List[int], List[str]]:
    if x:
        return [1, 1]
    else:
        return ['1', '1']

def g(x:bool):
    a,b = f(True)
```

Error:

```
z/test.py: note: In function "g":
z/test.py:10: error: 'Union[builtins.list[builtins.int], builtins.list[builtins.str]]' object is not iterable
```

### LINKED ISSUE #1855: error: 'Union[Any, Tuple[Any, Any]]' object is not iterable
Code:

``` python
from typing import *
Pair = Tuple[Any, Any]
def f(a: Union[Any, Pair]):
    x, y = a  # E: 'Union[Any, Tuple[Any, Any]]' object is not iterable
```

If the type of argument is either just `Any` or just `Pair` the code passes. But with the union it gives the error.

**UPDATE:** I don't think the unreduced union is at fault -- an example with two different Tuple types of length 2 also fails. It's more that the unpack operation isn't tried for each variant of the union.

### PR DESCRIPTION
**Pull request by elazarg**
Reopening #2154. Fix #1855: the union case was simply not handled (Probably fixes #1575 too).

I moved some code around so the cases are syntactically together. It makes the diff harder to read; some functions are easier to read without the diff. Most of the actual content is `check_multi_assign_from_union()`

The awkward part is the treatment of binding: I use a context manager to make the binding return a list of actions instead of doing the actual actions.

### CODE REVIEW

File: mypy/binder.py [hunk 1]
```diff
@@ -108,7 +110,7 @@ def push(self, node: Node, typ: Type) -> None:
     def get(self, expr: Union[Expression, Var]) -> Type:
         return self._get(expr.literal_hash)
 
-    def cleanse(self, expr: Expression) -> None:
+    def cleanse(self, expr: Node) -> None:
```
Comment (elazarg) [inline mypy/binder.py]: It's actually only used once, for NameExpr.

**Comment by elazarg** [comment]
@JukkaL I think I have addressed most of your comments from the original PR. Tell me if I miss anything. Also I am not confident with my "silence the binder" solution. I think it's ugly.

**Comment by gvanrossum** [comment]
@elazarg This has been languishing for a month now. Do you still want to work on this? Despite what GitHub says, it actually requires help rebasing (I tried and failed -- maybe because there are so many small local commits). Maybe you can restart this work by separating it in two parts, one that just moves code around as you wish without any changes in behavior (as proved by passing all tests without any test editing) and then a second one that actually makes the small(ish) changes needed to fix the issue.

**Comment by elazarg** [comment]
I will make it merge-able after #2238 is merged, since both deal with the binder.

**Comment by gvanrossum** [comment]
Since @2238 is now merged, can you update this too?

**Comment by elazarg** [comment]
I should, but it's pretty invasive and old so it will be tricky. And perhaps I should take a fresh look at the whole thing.

**Comment by ilevkivskyi** [comment]
@elazarg Are going to continue working on this? I think this is an important PR and I will be glad if it is finished.

**Comment by elazarg** [comment]
I tried, but the solutions I have in mind are too fundamental and will introduce too much churn. Sorry.

**Pull request closed without merging**